server/datastore/mysql/microsoft_mdm.go

```diff
@@ -1247,146 +1248,158 @@ func (ds *Datastore) ListMDMWindowsProfilesToInstall(ctx context.Context) ([]*fl
+//     and a NULL status. Other statuses mean that the operation is already in
+//     flight (pending), the operation has been completed but is still subject
+//     to independent verification by Fleet (verifying), or has reached a terminal
+//     state (failed or verified). If the profile's content is edited, all relevant hosts will
+//     be marked as status NULL so that it gets re-installed.
+//
+// Note that for label-based profiles, only fully-satisfied profiles are
+// considered for installation. This means that a broken label-based profile,
+// where one of the labels does not exist anymore, will not be considered for
+// installation.
+const windowsProfilesToInstallQuery = `
 	SELECT
 		ds.profile_uuid,
 		ds.host_uuid,
 		ds.name as profile_name
-	FROM ( %s ) as ds
+	FROM ( ` + windowsMDMProfilesDesiredStateQuery + ` ) as ds
 		LEFT JOIN host_mdm_windows_profiles hmwp
 			ON hmwp.profile_uuid = ds.profile_uuid AND hmwp.host_uuid = ds.host_uuid
 	WHERE
 		-- profiles in A but not in B
 		( hmwp.profile_uuid IS NULL AND hmwp.host_uuid IS NULL ) OR
 		-- profiles in A and B with operation type "install" and NULL status
 		( hmwp.host_uuid IS NOT NULL AND hmwp.operation_type = ? AND hmwp.status IS NULL )
-`, windowsMDMProfilesDesiredStateQuery)
+`
 
-	hostFilter := "TRUE"
-	if len(hostUUIDs) > 0 {
-		if len(onlyProfileUUIDs) > 0 {
-			hostFilter = "mwcp.profile_uuid IN (?) AND h.uuid IN (?)"
-		} else {
-			hostFilter = "h.uuid IN (?)"
-		}
+func (ds *Datastore) listAllMDMWindowsProfilesToInstallDB(ctx context.Context, tx sqlx.ExtContext) ([]*fleet.MDMWindowsProfilePayload, error) {
+	var profiles []*fleet.MDMWindowsProfilePayload
+	err := sqlx.SelectContext(ctx, tx, &profiles, fmt.Sprintf(windowsProfilesToInstallQuery, "TRUE", "TRUE", "TRUE", "TRUE"), fleet.MDMOperationTypeInstall)
+	if err != nil {
+		return nil, ctxerr.Wrapf(ctx, err, "selecting windows MDM profiles to install")
 	}
 
-	var err error
-	args := []any{fleet.MDMOperationTypeInstall}
-	query = fmt.Sprintf(query, hostFilter, hostFilter, hostFilter, hostFilter)
-	if len(hostUUIDs) > 0 {
+	return profiles, nil
+}
+
+func (ds *Datastore) listMDMWindowsProfilesToInstallDB(
+	ctx context.Context,
+	tx sqlx.ExtContext,
+	hostUUIDs []string,
+	onlyProfileUUIDs []string,
+) (profiles []*fleet.MDMWindowsProfilePayload, err error) {
+	if len(hostUUIDs) == 0 {
+		return profiles, nil
+	}
+
+	hostFilter := "h.uuid IN (?)"
+	if len(onlyProfileUUIDs) > 0 {
+		hostFilter = "mwcp.profile_uuid IN (?) AND h.uuid IN (?)"
+	}
+
+	toInstallQuery := fmt.Sprintf(windowsProfilesToInstallQuery, hostFilter, hostFilter, hostFilter, hostFilter)
+
+	// use a 10k host batch size to match what we do on the macOS side.
+	selectProfilesBatchSize := 10_000
+	if ds.testSelectMDMProfilesBatchSize > 0 {
+		selectProfilesBatchSize = ds.testSelectMDMProfilesBatchSize
+	}
+	selectProfilesTotalBatches := int(math.Ceil(float64(len(hostUUIDs)) / float64(selectProfilesBatchSize)))
```

Comment:
Victor added a generic batch-processing function to avoid implementing that subtle logic everywhere, if it's possible to refactor this to use it: https://github.com/fleetdm/fleet/blob/988927ceaa1a09d451a5f58ba5ffd0df44e11ea4/server/datastore/mysql/common_mysql/batch.go#L5

It is currently used only in one place, maybe we'll find it isn't quite general enough for all cases, but at a glance it looks like it could be used (`hostUUIDs` being the `payloads`, `selectProfilesBatchSize` the batch size), here is an example usage: https://github.com/fleetdm/fleet/blob/988927ceaa1a09d451a5f58ba5ffd0df44e11ea4/server/mdm/nanomdm/storage/mysql/queue.go#L297

Reply:
AHA I knew this existed somewhere!! thanks for the reminder.